stan/math/rev/functor/adj_jac_apply.hpp

```diff
@@ -118,381 +48,385 @@ struct compute_dims<Eigen::Matrix<T, R, C>> {
  * Eigen::Matrices with double or var scalar components
  */
 template <typename F, typename... Targs>
-struct adj_jac_vari : public vari {
-  std::array<bool, sizeof...(Targs)> is_var_;
-  using FReturnType
-      = std::result_of_t<F(decltype(is_var_), decltype(value_of(Targs()))...)>;
-
-  F f_;
-  std::array<int, sizeof...(Targs)> offsets_;
-  vari** x_vis_;
-  std::array<int, internal::compute_dims<FReturnType>::value> M_;
-  vari** y_vi_;
-
+class adj_jac_vari : public vari {
+ public:
+  static constexpr std::array<bool, sizeof...(Targs)> is_var_{
+      is_var<scalar_type_t<Targs>>::value...};
+
+ protected:
+  using FReturnType = plain_type_t<
+      std::result_of_t<F(decltype(value_of(std::declval<Targs>()))...)>>;
+  using x_vis_tuple_ = var_to_vari_filter_t<std::decay_t<Targs>...>;
+  F f_;  // Struct that with methods for computing forward and reverse pass.
+  x_vis_tuple_ x_vis_;  // tuple holding pointers to vari
   /**
-   * count_memory returns count (the first argument) + the number of varis used
-   * in the second argument + the number of arguments used to encode the
-   * variadic tail args.
-   *
-   * The adj_jac_vari constructor uses this to figure out how much space to
-   * allocate in x_vis_.
-   *
-   * The array offsets_ is populated with values to indicate where in x_vis_ the
-   * vari pointers for each argument will be stored.
-   *
-   * Each of the arguments can be an Eigen::Matrix with var or double scalar
-   * types, a std::vector with var, double, or int scalar types, or a var, a
-   * double, or an int.
-   *
-   * @tparam R number of rows, can be Eigen::Dynamic
-   * @tparam C number of columns, can be Eigen::Dynamic
-   * @tparam Pargs Types of rest of arguments
-   *
-   * @param count rolling count of number of varis that must be allocated
-   * @param x next argument to have its varis counted
-   * @param args the rest of the arguments (that will be iterated through
-   * recursively)
+   * Will be either vari* or vari**
    */
-  template <int R, int C, typename... Pargs>
-  inline size_t count_memory(size_t count, const Eigen::Matrix<var, R, C>& x,
-                             const Pargs&... args) {
-    static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
-    offsets_[t] = count;
-    count += x.size();
-    return count_memory(count, args...);
-  }
-
-  template <int R, int C, typename... Pargs>
-  inline size_t count_memory(size_t count, const Eigen::Matrix<double, R, C>& x,
-                             const Pargs&... args) {
-    static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
-    offsets_[t] = count;
-    return count_memory(count, args...);
+  typename internal::var_return_type<FReturnType>::type y_vi_;
+  /**
+   * Stores the dimensions of return value of forward pass to reconstruct
+   * container for reverse pass.
+   */
+  std::array<size_t, 2> dims_;
```

Comment:
So this is one of the weirder parts of `adj_jac_vari`. Because we can't keep an `Eigen::Matrix` or vector in `adj_jac_vari` (without making it inherit from `chainable_alloc`) we store the return of the forward pass in a `vari**`. To recreate the actual container that y returned for the reverse pass we need to keep the dimensions around. Can you think of a better way to handle this? Would be v interested!